File: CG/routeDiff/lagrangepolynomi.cpp
#include <vector>
#include <iostream>

class LagrangeAprx{
	public:
		LagrangeAprx(){}

		int getRequiredDataPoints(int degree, int inputOrder = 0){ return std::max(degree + 1, 2); }

		void interpolateOrderZero(
		        double x,
		        const std::vector<double> &xTable,
		        const std::vector<double> &yTable,
		        int yStride,
		        std::vector<double>& result
		) {
			int i;
			int j;
			int length = xTable.size();

			for(i = 0; i < yStride; i++){
				result.push_back(0);
			}

			for(i = 0; i < length; i++){
				double coefficient = 1;

				for(j = 0; j < length; j++){
				    if(j != i){
				        double diffX = xTable[i] - xTable[j];
				        coefficient *= (x - xTable[j]) / diffX;
				    }
				}

				for(j = 0;  j < yStride; j++){
				    result[j] += coefficient * yTable[i * yStride + j];
				}
			}
		}

		~LagrangeAprx(){}
};

int main(){

	double x = -67.72600000043167;

  std::vector<double> xTable;
	xTable.push_back(-225.0);
	xTable.push_back(-180.0);
	xTable.push_back(-135.0);
	xTable.push_back(-90.0);
	xTable.push_back(-45.0);
	xTable.push_back(-0.0);

  std::vector<double> yTable;
	yTable.push_back(-1944814.3726146347);
	yTable.push_back(-4779323.961685797);
	yTable.push_back(3740297.3814439983);
	yTable.push_back(-1946353.4244015086);
	yTable.push_back(-4779956.238836014);
	yTable.push_back(3738233.3529754514);
	yTable.push_back(-1945750.717271222);
	yTable.push_back(-4781652.001063456);
	yTable.push_back(3736280.714354399);
	yTable.push_back(-1943370.2943612777);
	yTable.push_back(-4783382.7269118745);
	yTable.push_back(3735607.1109777773);
	yTable.push_back(-1940506.515172479);
	yTable.push_back(-4783950.993914748);
	yTable.push_back(3736412.6206093826);
	yTable.push_back(-1938934.0602408552);
	yTable.push_back(-4783233.073741037);
	yTable.push_back(3738410.7113252394);

    int yStride = 3;
    std::vector<double> result;

	LagrangeAprx Lagrange;
	Lagrange.interpolateOrderZero(
		x,
		xTable,
		yTable,
		yStride,
		result
	);

	for (auto& index : result)
		std::cout<<"result"<<index<<std::endl;

	return 0;
}

File: CG/routeDiff/lagrangepolynomi.js
import defined from "./defined.js";

/**
 * An {@link InterpolationAlgorithm} for performing Lagrange interpolation.
 *
 * @namespace LagrangePolynomialApproximation
 */
var LagrangePolynomialApproximation = {
  type: "Lagrange",
};

/**
 * Given the desired degree, returns the number of data points required for interpolation.
 *
 * @param {Number} degree The desired degree of interpolation.
 * @returns {Number} The number of required data points needed for the desired degree of interpolation.
 */
LagrangePolynomialApproximation.getRequiredDataPoints = function (degree) {
  return Math.max(degree + 1.0, 2);
};

/**
 * Interpolates values using Lagrange Polynomial Approximation.
 *
 * @param {Number} x The independent variable for which the dependent variables will be interpolated.
 * @param {Number[]} xTable The array of independent variables to use to interpolate.  The values
 * in this array must be in increasing order and the same value must not occur twice in the array.
 * @param {Number[]} yTable The array of dependent variables to use to interpolate.  For a set of three
 * dependent values (p,q,w) at time 1 and time 2 this should be as follows: {p1, q1, w1, p2, q2, w2}.
 * @param {Number} yStride The number of dependent variable values in yTable corresponding to
 * each independent variable value in xTable.
 * @param {Number[]} [result] An existing array into which to store the result.
 * @returns {Number[]} The array of interpolated values, or the result parameter if one was provided.
 */
LagrangePolynomialApproximation.interpolateOrderZero = function (
  x,
  xTable,
  yTable,
  yStride,
  result
) {
  if (!defined(result)) {
    result = new Array(yStride);
  }

  var i;
  var j;
  var length = xTable.length;

  for (i = 0; i < yStride; i++) {
    result[i] = 0;
  }

  for (i = 0; i < length; i++) {
    var coefficient = 1;

    for (j = 0; j < length; j++) {
      if (j !== i) {
        var diffX = xTable[i] - xTable[j];
        coefficient *= (x - xTable[j]) / diffX;
      }
    }

    for (j = 0; j < yStride; j++) {
      result[j] += coefficient * yTable[i * yStride + j];
    }
  }

  return result;
};

// 案例
var x = -67.72600000043167;
var xTable = [
  -225.0,
  -180.0,
  -135.0,
  -90.0,
  -45.0,
  -0.0
];
var yTable = [
  -1944814.3726146347,
  -4779323.961685797,
  3740297.3814439983,
  -1946353.4244015086,
  -4779956.238836014,
  3738233.3529754514,
  -1945750.717271222,
  -4781652.001063456,
  3736280.714354399,
  -1943370.2943612777,
  -4783382.7269118745,
  3735607.1109777773,
  -1940506.515172479,
  -4783950.993914748,
  3736412.6206093826,
  -1938934.0602408552,
  -4783233.073741037,
  3738410.7113252394
];
var yStride = 3;
var result = [];

LagrangePolynomialApproximation.interpolateOrderZero(
	x,
	xTable,
	yTable,
	yStride,
	result
);

for (var i = 0; i < result.length; i++){
  console.log(`result的第 ${i} 个元素的值是:  ${result[i]}`);
}
